pkg/auth/handler/webapp/authflowv2/reset_password.go

```diff
@@ -75,10 +110,87 @@ func (h *AuthflowV2ResetPasswordHandler) GetErrorData(w http.ResponseWriter, r *
 }
 
 func (h *AuthflowV2ResetPasswordHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if isURLFromAdminAPI(r) {
+		h.serveHTTPNonAuthflow(w, r)
+	} else {
+		h.serveHTTPAuthflow(w, r)
+	}
+}
+
+func (h *AuthflowV2ResetPasswordHandler) serveHTTPNonAuthflow(w http.ResponseWriter, r *http.Request) {
+	code := r.URL.Query().Get("code")
+	makeHTTPHandler := func(handler func(w http.ResponseWriter, r *http.Request) error) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var err error
+			if code == "" {
+				err = newCodeRequiredErr()
+			} else {
+				err = handler(w, r)
+			}
+			if err != nil {
+				if apierrors.IsAPIError(err) {
+					// Still uses AuthflowController to render error because of same logic
+					h.Controller.ErrorRenderer.RenderError(w, r, err)
+				} else {
+					panic(err)
+				}
+			}
+		})
+	}
+
+	getHandler := makeHTTPHandler(func(w http.ResponseWriter, r *http.Request) error {
+		data, err := h.GetNonAuthflowData(w, r)
+		if err != nil {
+			return err
+		}
+		h.Renderer.RenderHTML(w, r, TemplateWebAuthflowResetPasswordHTML, data)
+		return nil
+	})
+
+	postHandler := makeHTTPHandler(func(w http.ResponseWriter, r *http.Request) error {
+		err := AuthflowResetPasswordSchema.Validator().ValidateValue(handlerwebapp.FormToJSON(r.Form))
+		if err != nil {
+			return err
+		}
+		newPassword := r.Form.Get("x_password")
+		confirmPassword := r.Form.Get("x_confirm_password")
+		err = pwd.ConfirmPassword(newPassword, confirmPassword)
+		if err != nil {
+			return err
+		}
+
+		err = h.Database.WithTx(func() error {
```

Comment:
@tung2744 Would love your view on this 🙏 In this `admin-api only reset password screen`, I encountered `hook-handle: transaction not started` error and fixed it by injecting a database handle to this webapp handler. 

Do you see any potential problems with this? Mine: Maybe too many concurrent database txns if many user reset password through admin api at once -> huge load to db?

Maybe I should inject the db handle to a higher level service? 🤔 

Thanks :adore:

Reply:
I think starting the transaction at handler level is correct. And I think it is already the highest level.

Reply:
I see i see thanks!